### DISCUSSION TITLE
Metrics represent by grouping

### DISCUSSION THREAD

**Discussion by creeew** [apache/skywalking#8127] [Q&A] [answered]
Since V8.3.0 service supproted group by service name. But there's only work on the service selection
![image](https://user-images.githubusercontent.com/30746864/141678381-ea544775-8ff5-4b01-b88a-30b007d0ad12.png)

And any plan for grouping these metrics value? Just like the picture above, the slow services chart would be return datas in the selected group "bookinfo"  only

If there's no plan to grouping metrics, would you be so kind to tell me some clues to solve this problem?

I've read the source code, find metrics's classes are generated by ftl and store in some metrics indices, like metrics-longavg/metrics-apex/metrics-cpm/metrics-percent and different metrics have different metric_table name in those indices like xx-cpm/xx-resp-time.
Here's what I'm doing:
(DB metrics by example)
1. rewrite the sourceBuilder toXXScope codes when receiving the agent report segment, in this we I can acquire the service name cause db scope got nothing relation with service. But I can get the server name by sourceServiceName
2. rewrite DatabaseXXMetrics, add a serviceName field
3. rewrite DatabaseXXMetricsBuilder#entity2Storage, get the group name by service name, and put the group as a index field and then we can query metrics by group
4. rewrite query method, like readMetricsValues and sortMetrics and readLabeledMetricsValues. Adding a new query param for group
It's kind of hard for me, I cannot tell how "ALL" scope metrics calcute if we grouping service metrics.

Thanks for considering this question. 
Best wishes.

**Comment by wu-sheng** [comment, marked as answer]
I think you don't have to follow the ways you proposed. Try to understand the OAL from higher level. What you asked is mostly asking for a new source. Read `org.apache.skywalking.oap.server.core.source.Source` and its implementations, including how they work with OAL engine, you may get my suggestions.
This would not be easy, it requires you fully understand the SkyWalking OAL engine.

Officially, I don't know anyone thinks this is meaningful, and plans to do this. The reason is very simple, SkyWalking is an APM, which focuses on performance, operation, devops, SRE fields. What you asked about are about BI or an Analysis system, a lot of tech stacks are super good and much better than SkyWalking to do this.

**Reply by creeew** [reply]
Thanks for your answer.
I've read the Source implementations. And these implementations like All/Service obtain the basic infomation from segment in MultiScopesAnalysisListener, and then pass basic information to specific metrics. Metrics can use OAL to do some filter or aggregation calcute. Maybe I can rewrite core.oal to filter metrics by group. But I still have to rewrite storageBuilder for storing the group filed in index.
I found a inspiring code when I read source code, `org.apache.skywalking.oap.server.core.analysis.data.MergableBufferedData#accept` will merge same metric's data if they got the same id(entity_id + timebucket) But I need all service metrics data and all service metrics grouped. Maybe there is nothing easy to do so ;(

As an APM system we may receive a lot of segments and these segments may come from different cluster. And I think I can classify different cluster metrics by service group. Just like topology only return the same group nodes.

**Reply by wu-sheng** [reply]
If you look closer, you should have noticed, group is never a concept, it is just a filter, and group is only a part of service full name. 

Topology by groups is also another filter mechanism, nothing more.

In the core concept document, https://skywalking.apache.org/docs/main/latest/en/concepts-and-designs/overview/#why-use-skywalking, it is very clear about these points. 
Of course, you could change anything you like in your fork, which has been licensed from Apache 2.0. I just clarify from project real current status.
